src/main/java/io/cdap/plugin/http/sink/batch/MessageBuffer.java

```diff
@@ -0,0 +1,203 @@
+ * Once the batch size is reached, the records are converted to the appropriate format and appended to the message.
+ * The message is then returned to the HTTPRecordWriter.
+ */
+public class MessageBuffer {
+  private static final String REGEX_HASHED_VAR = "#s*(\\w+)";
+  private final List<StructuredRecord> buffer;
+  private final String jsonBatchKey;
+  private final Boolean shouldWriteJsonAsArray;
+  private final String delimiterForMessages;
+  private final String charset;
+  private final String customMessageBody;
+  private final Function<List<StructuredRecord>, String> messageFormatter;
+  private final String contentType;
+  private static final Map<String, String> contentTypeMap = new HashMap<String, String>() {{
+    put("json", "application/json");
+    put("form", "application/x-www-form-urlencoded");
+    put("custom", "text/plain");
+  }};
+
+  public MessageBuffer(
+          String messageFormat, String jsonBatchKey, Boolean shouldWriteJsonAsArray,
+          String delimiterForMessages, String charset, String customMessageBody
+  ) {
+    this.jsonBatchKey = jsonBatchKey;
+    this.delimiterForMessages = delimiterForMessages;
+    this.charset = charset;
+    this.shouldWriteJsonAsArray = shouldWriteJsonAsArray;
+    this.customMessageBody = customMessageBody;
+    this.buffer = new ArrayList<>();
+    switch (messageFormat.toLowerCase()) {
+      case "json":
+        messageFormatter = this::formatAsJsonNoException;
+        contentType = contentTypeMap.get("json");
+        break;
+      case "form":
+        messageFormatter = this::formatAsForm;
+        contentType = contentTypeMap.get("form");
+        break;
+      case "custom":
+        messageFormatter = this::formatAsCustom;
+        contentType = contentTypeMap.get("custom");
+        break;
+      default:
+        throw new IllegalArgumentException("Invalid message format: " + messageFormat);
+    }
+  }
+
+  public void add(StructuredRecord record) {
+    buffer.add(record);
+  }
+
+  public void clear() {
+    buffer.clear();
+  }
+
+  public int size() {
+    return buffer.size();
+  }
+
+  public boolean isEmpty() {
+    return buffer.isEmpty();
+  }
+
+  public String getContentType() {
+    return contentType;
+  }
+
+  public String getMessage() throws IOException {
+    return messageFormatter.apply(buffer);
+  }
+
+  private String formatAsJsonNoException(List<StructuredRecord> buffer) {
+    try {
+      return formatAsJson(buffer);
+    } catch (IOException e) {
+      throw new IllegalStateException("Error formatting JSON message. Reason: " + e.getMessage(), e);
+    }
+  }
+
+  private String formatAsJson(List<StructuredRecord> buffer) throws IOException {
+    boolean useJsonBatchKey = !Strings.isNullOrEmpty(jsonBatchKey);
+    if (!shouldWriteJsonAsArray) {
+      return String.join(delimiterForMessages, getBufferAsJsonList());
+    }
+    if (useJsonBatchKey) {
```

Comment:
nit: more readable to put the short conditions at the start and then remove nesting for the larger code block:

```
if (!shouldWriteJsonAsArray) {
  return ...
}
if (!useJsonBatchKey) {
  return ...
} 
Schema bufferRecordSchema = ..
...
```

Reply:
Updated